editor/components/block-styles/index.js

```diff
@@ -0,0 +1,155 @@
+/**
+ * External dependencies
+ */
+import { find, compact, get } from 'lodash';
+import classnames from 'classnames';
+
+/**
+ * WordPress dependencies
+ */
+import { compose } from '@wordpress/element';
+import { withSelect, withDispatch } from '@wordpress/data';
+import { getBlockType } from '@wordpress/blocks';
+import { __, sprintf } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import './style.scss';
+import { BlockPreviewContent } from '../block-preview';
+
+/**
+ * Returns the active style from the given className.
+ *
+ * @param {Array} styles Block style variations.
+ * @param {string} className  Class name
+ *
+ * @return {Object?} The active style.
+ */
+export function getActiveStyle( styles, className ) {
+	let activeStyle;
+	className
+		.split( ' ' )
+		.map( ( current ) => current.trim() )
+		.filter( ( current ) => current.indexOf( 'is-style-' ) === 0 )
+		.forEach( ( current ) => {
+			if ( activeStyle ) {
+				return;
+			}
+			const potentialStyleName = current.substring( 9 );
+			activeStyle = find( styles, ( style ) => style.name === potentialStyleName );
+		} );
+	if ( ! activeStyle ) {
+		activeStyle = find( styles, ( style ) => style.isDefault );
+	}
+
+	return activeStyle;
+}
+
+/**
+ * Replaces the active style in the block's className.
+ *
+ * @param {string}  className   Class name.
+ * @param {Object?} activeStyle The replaced style.
+ * @param {Object}  newStyle    The replacing style.
+ *
+ * @return {string} The updated className.
+ */
+export function replaceActiveStyle( className, activeStyle, newStyle ) {
+	let added = false;
+	let updatedClassName = compact( className
+		.split( ' ' )
+		.map( ( current ) => {
+			const trimmed = current.trim();
+			if ( activeStyle && trimmed === 'is-style-' + activeStyle.name ) {
+				added = true;
+				return newStyle.isDefault ? null : 'is-style-' + newStyle.name;
+			}
+			return trimmed;
+		} ) ).join( ' ' );
+	if ( ! added && ! newStyle.isDefault ) {
+		updatedClassName = updatedClassName + ' is-style-' + newStyle.name;
+	}
+
+	return updatedClassName;
+}
+
+function BlockStyles( {
+	styles,
+	className,
+	onChangeClassName,
+	name,
+	attributes,
+	onSwitch,
+	onHoverClassName,
+} ) {
+	if ( ! styles ) {
+		return null;
+	}
+
+	const activeStyle = getActiveStyle( styles, className );
+	function updateClassName( style ) {
+		const updatedClassName = replaceActiveStyle( className, activeStyle, style );
+		onChangeClassName( updatedClassName );
+		onSwitch();
+	}
+
+	return (
+		<div className="editor-block-styles">
+			{ styles.map( ( style ) => {
+				const styleClassName = replaceActiveStyle( className, activeStyle, style );
+				/* eslint-disable jsx-a11y/click-events-have-key-events */
```

Comment:
This is an unfortunate instance of blindly disabling a rule for which its entire purpose is to protect against the thing we've introduced here.

https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/click-events-have-key-events.md